src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -44,11 +44,10 @@ public override async ValueTask<T> GetAssetAsync<T>(Checksum checksum, Cancellat
 
             using (Logger.LogBlock(FunctionId.AssetService_GetAssetAsync, Checksum.GetChecksumLogInfo, checksum, cancellationToken))
             {
-                // TODO: what happen if service doesn't come back. timeout?
                 var value = await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
 
-                _assetCache.TryAddAsset(checksum, value);
-                return (T)value;
+                // Attempt to add the value to the cache.  If someone else beat us, we'll return their value.
+                return (T)_assetCache.GetOrAdd(checksum, value);
```

Comment:
this also addresses a potential issue where two people might ask for a SerializableSourceText and get *different* instances of it.  This way, as long as the item is in the cache, you get the same instance (which can then also cache its SourceText for whover asks for it).